from launch import LaunchDescription
from launch.actions import DeclareLaunchArgument
from launch_ros.actions import Node


def generate_launch_description():

    # === LiDAR cluster publisher (detects wall + object) ===
    lidar_cluster = Node(
        package='lidar_detection',
        executable='lidar_cluster_publisher',
        name='lidar_cluster_publisher',
        output='screen',
        parameters=[{
            'gap_threshold': 0.2,
            'min_cluster_points': 1,
            'wall_length_threshold': 2.0,
            'wall_linearity_threshold': 0.001,
            'wall_min_points': 20,
            'object_length_threshold': 1.0,
            'object_max_points': 20,
            'max_range_ratio': 1.0,
            'obj_len_max': 1.0, #1.0
            'wal_len_min': 2.0, 
            'wal_lin_max': 0.001,
            'obj_nmp_min': 1,
            'wal_nmp_min': 150,
        }]
    )

    # === Object goal selector (filters + stabilizes clusters) ===
    object_selector = Node(
        package='lidar_detection',
        executable='object_goal_selector',
        name='object_goal_selector',
        output='screen',
        parameters=[{
            'cluster_distance_threshold': 2.5,
            'min_cluster_points': 8,
            'wall_thickness_threshold': 0.3,
            'stability_time': 3.0,
        }]
    )

    # === Goal sender (publishes visiting points & Nav2 goals) ===
    goal_sender = Node(
        package='lidar_detection',
        executable='send_goal_node',
        name='goal_sender',
        output='screen',
        parameters=[{
            'visit_offset': 0.75,
            'reach_threshold': 0.5,
        }]
    )

    return LaunchDescription([
        DeclareLaunchArgument(
            'use_sim_time',
            default_value='true',
            description='Use simulation (Gazebo) clock if true'),

        lidar_cluster,
        object_selector,
        goal_sender,
    ])
